membership-attribute-service/app/services/zuora/payment/ZuoraPaymentService.scala

```diff
@@ -113,61 +90,27 @@ class PaymentService(zuoraService: ZuoraSoapService, planMap: Map[ProductRatePla
+    zuoraService.previewInvoices(subId, numberOfBills).map { previewInvoiceItems =>
+      val maybeBillingSchedule = BillingSchedule.fromPreviewInvoiceItems(planMap, previewInvoiceItems)
+      maybeBillingSchedule.flatMap { billingSched =>
+        billingSched
+          .withCreditBalanceApplied(account.creditBalance)
+          .invoices
+          .list
+          .find(_.amount > 0)
+          .toOption
+      }
     }
-  }
-
-  def billingSchedule(subId: Id, accountId: AccountId, numberOfBills: Int)(implicit logPrefix: LogPrefix): Future[Option[BillingSchedule]] =
-    billingSchedule(subId, zuoraService.getAccount(accountId), numberOfBills)
 
-  def getPaymentMethod(accountId: AccountId, defaultMandateIdIfApplicable: Option[String] = None)(implicit
+  def getPaymentMethod(maybePaymentMethodId: Option[String], defaultMandateIdIfApplicable: Option[String] = None)(implicit
       logPrefix: LogPrefix,
   ): Future[Option[PaymentMethod]] =
-    getPaymentMethodByAccountId(accountId).map(_.flatMap(buildPaymentMethod(defaultMandateIdIfApplicable)))
-
-  private def getPaymentMethodByAccountId(accountId: AccountId)(implicit logPrefix: LogPrefix): Future[Option[Queries.PaymentMethod]] = {
-    val accountFuture = {
-      val account = zuoraService.getAccount(accountId)
-      account.onComplete {
-        case Failure(exception) =>
-          val message = scrub"Failed to get account for account $accountId"
-          logger.error(message, exception)
-        case Success(_) => logger.info(s"Successfully got account for $accountId")
-      }
-      account
-    }
-    def getPaymentMethod(account: Account) = {
-      val paymentMethod = getPaymentMethodByAccount(account)
-      paymentMethod.onComplete {
-        case Failure(exception) =>
-          val message = scrub"Failed to get payment method for account $accountId"
-          logger.error(message, exception)
-        case Success(_) => logger.info(s"Successfully got payment method for $accountId")
-      }
-      paymentMethod
-    }
-    for {
-      account <- accountFuture
-      paymentMethod <- getPaymentMethod(account)
-    } yield paymentMethod
-  }
-
-  private def getPaymentMethodByAccount(account: Account)(implicit logPrefix: LogPrefix): Future[Option[Queries.PaymentMethod]] = {
-    val maybeEventualPaymentMethod = account.defaultPaymentMethodId.map(zuoraService.getPaymentMethod)
-    maybeEventualPaymentMethod match {
-      case Some(eventualPaymentMethod) => eventualPaymentMethod.map(Some.apply)
-      case None => Future.successful(None)
-    }
-  }
+    (for {
+      paymentMethodId <- maybePaymentMethodId
+    } yield for {
+      soapPaymentMethod <- zuoraService.getPaymentMethod(paymentMethodId).withLogging(s"get payment method for $maybePaymentMethodId")
+    } yield buildPaymentMethod(defaultMandateIdIfApplicable, soapPaymentMethod))
+      .getOrElse(Future.successful(None))
```

Comment:
following on from @tjmw 's comment https://github.com/guardian/members-data-api/pull/1066#discussion_r1601598518
I realised there were 2 methods with similar name that were mostly just juggling the types badly, so I've turned it into this chained for comprehension that seems to avoid most of the fuss.